src/main/java/com/atlassian/db/replica/api/DualConnection.java

```diff
@@ -75,8 +107,13 @@ public CallableStatement prepareCall(String sql) throws SQLException {
     @Override
     public String nativeSQL(String sql) throws SQLException {
         checkClosed();
-        return connectionProvider.getWriteConnection(new RouteDecisionBuilder(Reason.RW_API_CALL).sql(sql)).nativeSQL(
-            sql);
+        if (experiments.isEnabled("com.atlassian.db.replica.experiment.nativesql.on.replica")) {
+            return connectionProvider.getReadConnection(new RouteDecisionBuilder(Reason.RO_API_CALL).sql(sql)).nativeSQL(
+                sql);
+        } else {
+            return connectionProvider.getWriteConnection(new RouteDecisionBuilder(Reason.RW_API_CALL).sql(sql)).nativeSQL(
```

Comment:
How do you see adding an experiment in ConnectionState? It looks like it will have to be injected along the way through all the classes in between, what require unproductive changes in unit tests.

StaticContext of Experiment would be a solution for it (it has it disadvantages)
I imaging that you would provide Experiment impl / configuration in DualConnectionBuilder as it is currently.

Then in any code in llb you could add:
`Experiement.getExperiment("this.is.just.example")`

Static code requires to have static mocks for configuration, what is bad, but doable.
On the other side, devloop is pretty fast and temporary changes in unit tests won't be a problem.

I don't have strong opinion, it looks like YAGNI.
Or maybe you had in mind that Experiments shouldn't be used in lol lvl internals?

Reply:
> It looks like it will have to be injected along the way through all the classes in between, what require unproductive changes in unit tests.

If injecting a new parameter will cause unit tests changes, then we have bad unit tests, and we should refactor them instead of avoiding changes.

> StaticContext of Experiment would be a solution for it

`static` is never a good choice. This approach would mask the problem with bad unit tests and will make them worse (harder to test, harder to run in parallel, possible flakiness).